File: Backend/routes/helper/gitOperations.ts
import { exec } from "child_process";
import { FileData } from "../../interface/fileData";
import * as fs from "fs";
import * as path from "path";
import { randomUUID } from "crypto";

export function isValidGitUrl(url: string): boolean {
  const gitUrlRegex = /^(git|https?):\/\/[^\s/$.?#].[^\s]*$/;
  return gitUrlRegex.test(url);
}

export async function executeGitClone(
  repoUrl: string,
  folderPath: string,
): Promise<void> {
  return new Promise((resolve, reject) => {
    exec(`git clone ${repoUrl} ${folderPath}`, (err, stdout, stderr) => {
      if (err) {
        console.error("Failed to clone:", stderr);
        reject(new Error("Failed to clone repository"));
      } else {
        resolve();
      }
    });
  });
}

File: go-backend/constants/gitOperations.go
package constants

import (
	"os"
	"regexp"

	"github.com/go-git/go-git/v6"
)

func IsValidGitUrl(url string) bool {
	gitUrlRegex := regexp.MustCompile(`^(git|https?):\/\/[^\s/$.?#].[^\s]*$`)
	return gitUrlRegex.MatchString(url)
}

func ExecuteGitClone(repoUrl string, folderPath string) bool {
	_, err := git.PlainClone(folderPath, &git.CloneOptions{
		URL:      repoUrl,
		Progress: os.Stdout,
	})
	if err != nil {
		return false
	}
	return true
}
